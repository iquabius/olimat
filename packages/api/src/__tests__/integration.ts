import { gql } from 'apollo-server-express';
import { createTestClient } from 'apollo-server-testing';
// https://www.prisma.io/forum/t/testing-with-graphql-request/537
import nock from 'nock';

import { createTestServer } from './__utils';

const getCitiesQuery = gql`
  query getCities {
    cities {
      id
      name
    }
  }
`;

const getCityQuery = gql`
  query getCity($id: ID!) {
    city(id: $id) {
      id
      name
    }
  }
`;

describe('Queries', () => {
  beforeAll(() => nock.disableNetConnect());
  afterAll(() => nock.enableNetConnect());

  afterEach(() => nock.cleanAll());

  test('fetches list of cities', async () => {
    const mockCities = [
      { id: 'cId1', name: 'Cáceres' },
      { id: 'cId2', name: 'Cuiabá' },
      { id: 'cId3', name: 'Sinop' },
    ];

    // Intercepta requisição ao servidor do prisma e simula uma resposta.
    // Use o 'recorder' pra inspecionar as requisições.
    // nock.recorder.rec();
    nock('http://localhost:4466')
      .post('/olimat-api/dev')
      .reply(200, [
        {
          data: {
            cities: mockCities,
          },
        },
      ]);

    const server = createTestServer();

    // Usa o servidor de teste como entrata para a função createTestClient.
    // Isso nos fornece uma interface parecida com o apolloClient.query,
    // para executar consultas a instância do ApolloServer.
    const { query } = createTestClient(server);
    const res = await query({ query: getCitiesQuery });
    expect(res).toMatchSnapshot();
  });

  // 'query', de createTestServer(), está com o tipo errado, parece um bug.
  test('fetches single city', async () => {
    const mockCity = { id: 'cId1', name: 'Barra do Bugres' };

    // nock.recorder.rec();
    nock('http://localhost:4466')
      .post('/olimat-api/dev')
      .reply(200, [
        {
          data: {
            city: mockCity,
          },
        },
      ]);

    const server = createTestServer();

    const { query } = createTestClient(server);
    const res = await query({ query: getCityQuery, variables: { id: mockCity.id } });
    expect(res).toMatchSnapshot();
    expect(res.data.city).toEqual(mockCity);
  });
});
