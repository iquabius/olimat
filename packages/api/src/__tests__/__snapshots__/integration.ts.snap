// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries fetches list of cities 1`] = `
Object {
  "data": Object {
    "cities": Array [
      Object {
        "id": "cId1",
        "name": "Cáceres",
      },
      Object {
        "id": "cId2",
        "name": "Cuiabá",
      },
      Object {
        "id": "cId3",
        "name": "Sinop",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
